### PR TITLE
[LOG-11302] Added UTs for SSL and Regions + removal of Logentries mentions

### PR DESCRIPTION
**Pull request by mzabad-r7**
Jira: https://issues.corp.rapid7.com/browse/LOG-11302

Notable changes in this PR:
 - UT checks for SSL flag set to `true`, `false` and `not set`.
 - UT checks for region flag set to `allowed regions`, `wrong region short name` and `not set`.
     - Now an error is thrown if the region is not supplied or the wrong region is given, as opposed to setting it to `undefined` and waiting for it to break down the line.
 - Replacement of `Logentries` mentions with `Rapid7` or `IOPS` where appropriate.
 - Version change for `Karma-PhantomJs` from `0.1.4` to `1.0.4`
     - Not quite sure how the UTs could pass in production without this, locally they failed until I upped the version.

I've discussed with PD to enable Travis for this project, no update from them yet.

FYI:  @bafanasiev-r7 @dharibabu-r7 @ashulga-r7

### CODE REVIEW

**Comment by kmackay-r7** [comment]
https://github.com/rapid7/r7insight_js/pull/3/files#diff-01a001b61a6def32199acd8a19618c1fR128

Does `var _endpoint = "js.logentries.com/v1";` on this line need to change as well?

**Comment by kmackay-r7** [comment]
https://github.com/rapid7/r7insight_js/blob/86add9975b5d62b6e5ff73fb5ba4af78ce4a6967/.travis.yml#L12

I know you mentioned travis in the comment already, but does `repo: logentries/le_js` need to change too?

**Comment by mzabad-r7** [comment]
> https://github.com/rapid7/r7insight_js/pull/3/files#diff-01a001b61a6def32199acd8a19618c1fR128

Missed that one. It's part of the `dyn` file so it should technically regenerate with the right url but for some reason it didn't.
Fixed now. Thanks :D 

As for Travis, not modifying that file yet as it will likely need to be completely changed if PD decides so.

File: src/le.js [hunk 1]
```diff
@@ -1,6 +1,6 @@
 /**
- * @license Copyright 2015 Logentries.
- * Please view license at https://raw.github.com/logentries/le_js/master/LICENSE
+ * @license Copyright 2019 Rapid7.
+ * Please view license at https://raw.github.com/rapid7/r7insight_js/master/LICENSE
```
Comment (kmackay-r7) [inline src/le.js]: Is this deliberately changed from `le_js` to `r7insight_js` ?

File: test/leSpec.js [hunk 2]
```diff
@@ -310,53 +321,155 @@ describe('catch all option', function () {
...
+    it('SSL option set to true leads to \"https\"', function () {
+        IOPS.init({
+            token: TOKEN,
+            name: 'test',
+            ssl: true,
+            region: 'eu'
+        });
+        IOPS.log("Test");
+        var url = this.requestList[0].url;
+        assert(url.search('https') === 0);
```
Comment (kmackay-r7) [inline test/leSpec.js]: `indexOf` would work fine here. No difference in this case, but worth being away that the parameter you pass to `string.search()` is always converted to a regex, so IMHO we shouldn't use it in unless we're actually doing a regex match

File: test/leSpec.js [hunk 3]
```diff
@@ -310,53 +321,155 @@ describe('catch all option', function () {
...
+    it('SSL option set to false leads to \"http\"', function () {
+        IOPS.init({
+            token: TOKEN,
+            name: 'test',
+            ssl: false,
+            region: 'eu'
+        });
+        IOPS.log("Test");
+        var url = this.requestList[0].url;
+        assert(url.search("http") === 0);
```
Comment (kmackay-r7) [inline test/leSpec.js]: This will pass even if the url begins with `https`; you'd want to check that the url begins with `http` and also check that the url _doesn't_ begin with `https` (or just check for `http:`)

A good rule of thumb for tests like this is to make them fail first, then make them pass (i.e. break the logic so that it returns `https`, observe the test failing, then fix the logic)

File: test/leSpec.js [hunk 4]
```diff
@@ -411,10 +525,24 @@ describe('print option', function () {
...
 
     afterEach(restoreXMLHttpRequests);
     afterEach(destroy);
 });
+
+function fail(message) {
+    throw message;
+}
+
+function assert(condition, message) {
```
Comment (kmackay-r7) [inline test/leSpec.js]: We're the using the [Jasmine](https://jasmine.github.io/) test framework in this project, so you can use its assertions (`expect(actual).toBe(expected)`).
It will give you nice messages when tests fail, and report those failures properly in CI builds (jenkins or whatever).
For custom messages, you can do `expect(actual).toBe(expected, 'my custom message')`

File: src/le.js (hunk 1 above)
Comment (mzabad-r7) [reply src/le.js]: Yes, it serves the License in the repo. The public one will be modified when it will be merged

File: test/leSpec.js (hunk 3 above)
Comment (mzabad-r7) [reply test/leSpec.js]: Very good point. I missed that.

File: bower.json [hunk 5]
```diff
@@ -1,8 +1,8 @@
 {
   "name": "le_js",
   "version": "0.0.3",
-  "homepage": "https://logentries.com",
-  "description": "Javascript library for Logentries.com",
+  "homepage": "https://rapid7.com",
+  "description": "Javascript library for rapid7.com",
```
Comment (rjacobs-r7) [inline bower.json]:
```suggestion
  "description": "Javascript library for Rapid7 Insight",
```

File: component.json [hunk 6]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "le_js",
-  "repo": "logentries/le_js",
-  "description": "CLient-side JavaScript logging library for Logentries",
+  "repo": "rapid7/le_js",
+  "description": "Client-side JavaScript logging library for Rapid7",
```
Comment (rjacobs-r7) [inline component.json]:
```suggestion
  "description": "Client-side JavaScript logging library for Rapid7 Insight",
```

File: bower.json (hunk 5 above)
Comment (rjacobs-r7) [reply bower.json]: same in other places

File: package.json [hunk 7]
```diff
@@ -13,14 +13,14 @@
   },
   "repository": {
     "type": "git",
-    "url": "git://github.com/logentries/le_js.git"
+    "url": "git://github.com/rapid7/le_js.git"
```
Comment (rjacobs-r7) [inline package.json]: wrong repo url

File: product/le.dyn.js [hunk 8]
```diff
@@ -125,7 +125,7 @@ var XMLHttpRequest = function() {
         /**
          * @const
          * @type {string} */
-        var _endpoint = "js.logentries.com/v1";
+        var _endpoint = "js.logs.insight.rapid7.com/v1";
```
Comment (rjacobs-r7) [inline product/le.dyn.js]: Did you verify that this endpoint exists and works? Because I'm not entirely sure it does...

File: product/le.min.js [hunk 9]
```diff
@@ -1,13 +1,13 @@
 /*
```
Comment (rjacobs-r7) [inline product/le.min.js]: Can I assume this whole file was re-generated?

File: src/le.js [hunk 10]
```diff
@@ -409,4 +409,21 @@
                 loggers[k].info.apply(this, arguments);
         }
     };
+
+    function validate_region (region) {
```
Comment (rjacobs-r7) [inline src/le.js]:
```suggestion
    function validate_region(region) {
```

File: src/le.js [hunk 11]
```diff
@@ -409,4 +409,21 @@
...
+        var allowed_regions = ['eu', 'us', 'ca', 'au', 'ap'];
+        if (region) {
+            if (allowed_regions.indexOf(region) > -1) {
+                return region;
+            } else {
```
Comment (rjacobs-r7) [inline src/le.js]: Already returning above. `else` isn't needed :)

File: src/le.js [hunk 12]
```diff
@@ -409,4 +409,21 @@
                 loggers[k].info.apply(this, arguments);
         }
     };
+
+    function validate_region (region) {
+        var allowed_regions = ['eu', 'us', 'ca', 'au', 'ap'];
+        if (region) {
+            if (allowed_regions.indexOf(region) > -1) {
```
Comment (rjacobs-r7) [inline src/le.js]: These ancient javascript versions make me want to cry

File: product/le.min.js (hunk 9 above)
Comment (mzabad-r7) [reply product/le.min.js]: It should've been.
The only way it could've been affected would be if refactoring edited it.

File: product/le.dyn.js (hunk 8 above)
Comment (mzabad-r7) [reply product/le.dyn.js]: It comes from the code found in `product/le.js` 
https://github.com/rapid7/r7insight_js/blob/0b3a79fca9dc402767bb6be2ca34b1f16cc05051/product/le.js#L89

**Comment by mzabad-r7** [comment]
I ranamed the files too to not include `le` in the name.
The project still builds fine and the tests all pass.

File: README.md [hunk 13]
```diff
@@ -1,4 +1,4 @@
-le.js
+r7insight_js.js
```
Comment (rjacobs-r7) [inline README.md]:
```suggestion
r7insight.js
```

File: bower.json [hunk 14]
```diff
@@ -1,9 +1,9 @@
 {
-  "name": "le_js",
+  "name": "r7insight_js",
   "version": "0.0.3",
```
Comment (rjacobs-r7) [inline bower.json]: We probably want to bump to (to 1.0?)

File: src/r7insight_js.js [hunk 15]
```diff
@@ -410,20 +410,19 @@
...
                 throw_error("Unrecognised region");
             }
-        } else {
-            throw_error("No region defined");
         }
+        throw_error("No region defined");
     }
 
-    function throw_error (message) {
+    function throw_error(message) {
```
Comment (rjacobs-r7) [inline src/r7insight_js.js]: Is this really useful? 😂

Comment (mzabad-r7) [reply src/r7insight_js.js]: Yes, it helps you avoid writing a function for every single throw you want to do.
It looks stupid, I know, but it helps for nicer code.

Comment (rjacobs-r7) [reply src/r7insight_js.js]: ... Why dop you have to write a function for throws? :p All you're doing is just `throw "message"`

Comment (mzabad-r7) [reply src/r7insight_js.js]: Ok so, since I separated the `validate_region` into a function then the `throw` lives inside a function, which is great.
Initially, it wasn't in a function and it gave out.

TLDR;
I removed the `throw_error` function cause it was no longer needed.

File: README.md (hunk 13 above)
Comment (mzabad-r7) [reply README.md]: y u do dis :(

**Comment by mzabad-r7** [comment]
Travis seems to work, but the build seems to fail with a `0 exit code`.
I believe that it's failing because of `saucelabs`

File: product/le.dyn.js (hunk 8 above)
Comment (rjacobs-r7) [reply product/le.dyn.js]: ... and you've tested that this works? :P Also how does it know which region to send log files to?

File: .travis.yml [hunk 16]
```diff
@@ -1,15 +1,15 @@
 language: node_js
 node_js:
-- 0.1
+- 8.10.0
```
Comment (rjacobs-r7) [inline .travis.yml]:
```suggestion
- lts/carbon
```

File: .travis.yml [hunk 17]
```diff
@@ -1,15 +1,15 @@
...
 sudo: false
 deploy:
   provider: releases
   api_key:
     secure: S2JVfYhWj4b2DvjsPtJYqUrAUtC9Mwdc1kMkA+p/LeSYJOZL8gFq55YReAPI1pNKSTJxwnxdlbMb5gtxyItVj7vz23Inv5S1DwialwgTff3ZV1JOquT91BeHnvphjBdm5L8G65QI35xEu5d1AuFr+cv6q+ocGr/k70b56fnEPcs=
```
Comment (rjacobs-r7) [inline .travis.yml]: What's this key for? If it's a Saucelabs API key we can get rid of it.

File: product/le.dyn.js (hunk 8 above)
Comment (mzabad-r7) [reply product/le.dyn.js]: Not quite sure why, but `_endpoint` gets another value on the next line 
image
